# Experiments

In this notebook, we experiment and evaluate different configurations,
in order to find one which gives the best performance.
This includes tuning hyperparamaters, and trying various data augmentation strategies.

Across all configurations, we are attempting to finetune a RoBERTa model.

In [1]:
!pip install --quiet datasets
from modeleval import *

## Baseline
We first fine-tune RoBERTa on with no data augmentation strategy, for $3$ epochs, at a learning rate of $3 \times 10 ^{-5}$. We use a linear learning rate scheduler, and a batch size of $8$.

In [ ]:
evaluate_model()

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Training on CUDA
Training fold 1...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 1 evaluation results: {'eval_f1': 0.547945205479452, 'eval_loss': 0.35339587926864624, 'eval_runtime': 11.9286, 'eval_samples_per_second': 140.419, 'eval_steps_per_second': 17.605, 'epoch': 3.0}
Training fold 2...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 2 evaluation results: {'eval_f1': 0.5531914893617021, 'eval_loss': 0.29567497968673706, 'eval_runtime': 11.7222, 'eval_samples_per_second': 142.891, 'eval_steps_per_second': 17.915, 'epoch': 3.0}
Training fold 3...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 3 evaluation results: {'eval_f1': 0.535593220338983, 'eval_loss': 0.3302139639854431, 'eval_runtime': 11.7422, 'eval_samples_per_second': 142.648, 'eval_steps_per_second': 17.884, 'epoch': 3.0}
Training fold 4...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 4 evaluation results: {'eval_f1': 0.0, 'eval_loss': 0.31710806488990784, 'eval_runtime': 11.7547, 'eval_samples_per_second': 142.496, 'eval_steps_per_second': 17.865, 'epoch': 3.0}
Training fold 5...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 5 evaluation results: {'eval_f1': 0.41379310344827586, 'eval_loss': 0.2682809829711914, 'eval_runtime': 11.7165, 'eval_samples_per_second': 142.961, 'eval_steps_per_second': 17.923, 'epoch': 3.0}
Average evaluation results from cross-validation: {'eval_f1': 0.41010460372568264, 'eval_loss': 0.3129347741603851, 'eval_runtime': 11.772839999999999, 'eval_samples_per_second': 142.283, 'eval_steps_per_second': 17.8384, 'epoch': 3.0}


{'eval_f1': 0.41010460372568264,
 'eval_loss': 0.3129347741603851,
 'eval_runtime': 11.772839999999999,
 'eval_samples_per_second': 142.283,
 'eval_steps_per_second': 17.8384,
 'epoch': 3.0}

As a baseline, we find that simply fine-tuning RoBERTa on the data set yields an average F1 Score of $0.4138$. This is low, and is likely a result of the large imbalance between positive and negative samples -- there are roughly 10 times more negative samples than positive
samples.

## Simple Oversampling / Undersampling
In order to combat this label imbalance, we perform a naive oversampling / undersampling of the dataset. We resample the minority class five times, and undersample the majority class by a factor of two. This yields a balanced data set.

In [3]:
class NaiveResampling(AugmentationStrategy):

    def __init__(self, oversample_ratio: float, under_sample_ratio:float):
        self.oversample_ratio = oversample_ratio
        self.under_sample_ratio = under_sample_ratio

    def augment_data(self, data: Dataset) -> Dataset:
        df = data.to_pandas()
        df_pos = df[df['label'] == 1]
        df_neg = df[df['label'] == 0]
        df_pos = df_pos.sample(frac=self.oversample_ratio, replace=True)
        df_neg = df_neg.sample(frac=min(self.under_sample_ratio, 1.0), replace=False)
        df = pd.concat([df_pos, df_neg])
        return Dataset.from_pandas(df)

In [ ]:
evaluate_model(augmentation_strategy=NaiveResampling(oversample_ratio=5, under_sample_ratio=0.5))

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Training on CUDA
Training fold 1...
Counter({1: 3180, 0: 3032})


Map:   0%|          | 0/6212 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 6212
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 1 evaluation results: {'eval_f1': 0.5525525525525525, 'eval_loss': 0.604171872138977, 'eval_runtime': 11.9304, 'eval_samples_per_second': 140.398, 'eval_steps_per_second': 17.602, 'epoch': 3.0}
Training fold 2...
Counter({1: 3175, 0: 3032})


Map:   0%|          | 0/6207 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 6207
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 2 evaluation results: {'eval_f1': 0.5938375350140056, 'eval_loss': 0.5396053194999695, 'eval_runtime': 11.9522, 'eval_samples_per_second': 140.142, 'eval_steps_per_second': 17.57, 'epoch': 3.0}
Training fold 3...
Counter({1: 3175, 0: 3032})


Map:   0%|          | 0/6207 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 6207
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 3 evaluation results: {'eval_f1': 0.5738161559888579, 'eval_loss': 0.5643535256385803, 'eval_runtime': 11.8995, 'eval_samples_per_second': 140.762, 'eval_steps_per_second': 17.648, 'epoch': 3.0}
Training fold 4...
Counter({1: 3175, 0: 3032})


Map:   0%|          | 0/6207 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 6207
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 4 evaluation results: {'eval_f1': 0.5263157894736842, 'eval_loss': 0.5804007649421692, 'eval_runtime': 11.8748, 'eval_samples_per_second': 141.055, 'eval_steps_per_second': 17.685, 'epoch': 3.0}
Training fold 5...
Counter({1: 3175, 0: 3032})


Map:   0%|          | 0/6207 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 6207
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 5 evaluation results: {'eval_f1': 0.5329512893982808, 'eval_loss': 0.6596797704696655, 'eval_runtime': 11.8574, 'eval_samples_per_second': 141.262, 'eval_steps_per_second': 17.71, 'epoch': 3.0}
Average evaluation results from cross-validation: {'eval_f1': 0.5558946644854762, 'eval_loss': 0.5896422505378723, 'eval_runtime': 11.90286, 'eval_samples_per_second': 140.72379999999998, 'eval_steps_per_second': 17.643, 'epoch': 3.0}


{'eval_f1': 0.5558946644854762,
 'eval_loss': 0.5896422505378723,
 'eval_runtime': 11.90286,
 'eval_samples_per_second': 140.72379999999998,
 'eval_steps_per_second': 17.643,
 'epoch': 3.0}

This technique yielded an increased performance, with an average F1 score of $0.5559$ across the five folds.

Next we repeat the experiment, but we only use resampling.

In [4]:
evaluate_model(augmentation_strategy=NaiveResampling(oversample_ratio=10.0, under_sample_ratio=1.0))

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Training on CUDA
Training fold 1...
Counter({1: 6360, 0: 6064})


Map:   0%|          | 0/12424 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 12424
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:90: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 1 evaluation results: {'eval_f1': 0.526984126984127, 'eval_loss': 0.46870145201683044, 'eval_runtime': 11.6651, 'eval_samples_per_second': 143.591, 'eval_steps_per_second': 18.002, 'epoch': 3.0}
Training fold 2...
Counter({1: 6350, 0: 6065})


Map:   0%|          | 0/12415 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 12415
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:90: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 2 evaluation results: {'eval_f1': 0.5960264900662252, 'eval_loss': 0.5179898738861084, 'eval_runtime': 11.6396, 'eval_samples_per_second': 143.906, 'eval_steps_per_second': 18.042, 'epoch': 3.0}
Training fold 3...
Counter({1: 6350, 0: 6065})


Map:   0%|          | 0/12415 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 12415
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:90: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 3 evaluation results: {'eval_f1': 0.5614035087719298, 'eval_loss': 0.5985027551651001, 'eval_runtime': 11.6991, 'eval_samples_per_second': 143.173, 'eval_steps_per_second': 17.95, 'epoch': 3.0}
Training fold 4...
Counter({1: 6350, 0: 6065})


Map:   0%|          | 0/12415 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 12415
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:90: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 4 evaluation results: {'eval_f1': 0.509090909090909, 'eval_loss': 0.7149061560630798, 'eval_runtime': 11.8083, 'eval_samples_per_second': 141.849, 'eval_steps_per_second': 17.784, 'epoch': 3.0}
Training fold 5...
Counter({1: 6350, 0: 6065})


Map:   0%|          | 0/12415 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 12415
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:90: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 5 evaluation results: {'eval_f1': 0.5454545454545454, 'eval_loss': 0.40833359956741333, 'eval_runtime': 11.8147, 'eval_samples_per_second': 141.773, 'eval_steps_per_second': 17.775, 'epoch': 3.0}
Average evaluation results from cross-validation: {'eval_f1': 0.5477919160735473, 'eval_loss': 0.5416867673397064, 'eval_runtime': 11.72536, 'eval_samples_per_second': 142.85840000000002, 'eval_steps_per_second': 17.9106, 'epoch': 3.0}


{'eval_f1': 0.5477919160735473,
 'eval_loss': 0.5416867673397064,
 'eval_runtime': 11.72536,
 'eval_samples_per_second': 142.85840000000002,
 'eval_steps_per_second': 17.9106,
 'epoch': 3.0}

This gives us an F1 score of $0.5478$

## Synonym Replacement

Next we attempt a method of data augmentation wherein words are probabilistically replaced with synonyms.

Approach:
1. Repeat `num_augmented_versions` times:
2. For each sample in the minority class:
3. Replace each word probabilistically with a chance of `synonym_ratio`.
4. Concat onto train dataset.

In [2]:
import random
import nltk
from nltk.corpus import wordnet
from typing import List
import pandas as pd

class SynonymReplacement(AugmentationStrategy):
    def __init__(self, synonym_ratio: float, num_augmented_versions: int = 1):
        self.synonym_ratio = synonym_ratio
        self.num_augmented_versions = num_augmented_versions
        nltk.download('wordnet')
        nltk.download('omw-1.4')

    def get_synonyms(self, word: str) -> List[str]:
        synonyms = set()
        for syn in wordnet.synsets(word):
            for lemma in syn.lemmas():
                synonym = lemma.name().replace('_', ' ')
                if synonym.lower() != word.lower():
                    synonyms.add(synonym)
        return list(synonyms)

    def replace_synonyms(self, text: str) -> str:
        words = text.split()
        num_to_replace = max(1, int(len(words) * self.synonym_ratio))
        words_to_replace = random.sample(words, num_to_replace)

        for i, word in enumerate(words):
            if word in words_to_replace:
                synonyms = self.get_synonyms(word)
                if synonyms:
                    words[i] = random.choice(synonyms)

        return ' '.join(words)

    def augment_data(self, data: Dataset) -> Dataset:
        df = data.to_pandas()
        df_pos = df[df['label'] == 1]

        for _ in range(self.num_augmented_versions):
            temp = df_pos.copy()
            temp['text'] = temp['text'].apply(self.replace_synonyms)
            df = pd.concat([df, temp])

        return Dataset.from_pandas(df)

In [5]:
evaluate_model(augmentation_strategy=SynonymReplacement(synonym_ratio=0.1, num_augmented_versions=9))

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Training on CUDA
Training fold 1...
Counter({1: 7940, 0: 7581})


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 1 evaluation results: {'eval_f1': 0.5393794749403341, 'eval_loss': 0.2463763952255249, 'eval_runtime': 58.6716, 'eval_samples_per_second': 142.744, 'eval_steps_per_second': 17.845, 'epoch': 3.0}
Training fold 2...
Counter({1: 7940, 0: 7581})


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 2 evaluation results: {'eval_f1': 0.0, 'eval_loss': 0.31363847851753235, 'eval_runtime': 58.5157, 'eval_samples_per_second': 143.124, 'eval_steps_per_second': 17.893, 'epoch': 3.0}
Training fold 3...
Counter({1: 7940, 0: 7581})


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 3 evaluation results: {'eval_f1': 0.39466666666666667, 'eval_loss': 0.28392454981803894, 'eval_runtime': 58.9524, 'eval_samples_per_second': 142.064, 'eval_steps_per_second': 17.76, 'epoch': 3.0}
Training fold 4...
Counter({1: 7940, 0: 7581})


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 4 evaluation results: {'eval_f1': 0.0, 'eval_loss': 0.3018490970134735, 'eval_runtime': 58.8692, 'eval_samples_per_second': 142.265, 'eval_steps_per_second': 17.785, 'epoch': 3.0}
Training fold 5...
Counter({1: 7940, 0: 7581})


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 5 evaluation results: {'eval_f1': 0.0, 'eval_loss': 0.3139241635799408, 'eval_runtime': 58.1574, 'eval_samples_per_second': 144.006, 'eval_steps_per_second': 18.003, 'epoch': 3.0}
Average evaluation results from cross-validation: {'eval_f1': 0.18680922832140015, 'eval_loss': 0.2919425368309021, 'eval_runtime': 58.63326000000001, 'eval_samples_per_second': 142.8406, 'eval_steps_per_second': 17.8572, 'epoch': 3.0}


{'eval_f1': 0.18680922832140015,
 'eval_loss': 0.2919425368309021,
 'eval_runtime': 58.63326000000001,
 'eval_samples_per_second': 142.8406,
 'eval_steps_per_second': 17.8572,
 'epoch': 3.0}

In [18]:
class LlmRewording(AugmentationStrategy):
    def __init__(self, resampler: NaiveResampling, num_augmented_versions: int = 1):
        self.resampler = resampler
        self.num_augmented_versions = num_augmented_versions

    def augment_data(self, data: Dataset) -> Dataset:
        df = data.to_pandas()
        df_pos = df[df['label'] == 1]
        df_reworded = pd.read_csv("data/reworded.csv")
        df_reworded["text"] = df_reworded["reworded"]
        df_reworded = df_reworded.drop("reworded", axis=1)
        df_reworded = df_reworded[df_reworded["par_id"].isin(df_pos["par_id"])]
        df = pd.concat([df, df_reworded])
        df = df.reset_index(drop=True)
        return self.resampler.augment_data(Dataset.from_pandas(df))

In [19]:
evaluate_model(
    augmentation_strategy=LlmRewording(
        resampler=NaiveResampling(oversample_ratio=5.0, under_sample_ratio=1.0),
        num_augmented_versions=1
    )
)

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Training on CUDA
Training fold 1...
Counter({1: 7835, 0: 7581})


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


KeyboardInterrupt: 